File: Homework1/Hw1/Parser.cs
namespace Hw1;

public static class Parser
{
    public static void ParseCalcArguments(string[] args,
        out double val1,
        out CalculatorOperation operation,
        out double val2)
    {
        if (!IsArgLengthSupported(args))
            throw new ArgumentException("Invalid number of arguments specified");
        val1 = ParseStringToDouble(args[0]);
        operation = ParseOperation(args[1]);
        val2 = ParseStringToDouble(args[2]);
        }

    private static double ParseStringToDouble(string s)
    {
        if (double.TryParse(s, out var number))
        {
            return number;
        }
        throw new ArgumentException("The entered argument is not a number");
    }

    private static bool IsArgLengthSupported(string[] args) => args.Length == 3;

    private static CalculatorOperation ParseOperation(string arg)
    {
        return arg switch
        {
            "+" => CalculatorOperation.Plus,
            "-" => CalculatorOperation.Minus,
            "*" => CalculatorOperation.Multiply,
            "/" => CalculatorOperation.Divide,
            _ => throw new InvalidOperationException("Operation specified incorrectly")
        };
    }
}

File: Homework4/Hw4/Parser.fs
module Hw4.Parser
open Hw4.Calculator

type CalcOptions = {
    arg1: float
    arg2: float
    operation: CalculatorOperation
}

let isArgLengthSupported (args : string[]) = args.Length = 3
    
let parseOperation (arg : string) =
    match arg with
    | "+" -> CalculatorOperation.Plus
    | "-" -> CalculatorOperation.Minus
    | "*" -> CalculatorOperation.Multiply
    | "/" -> CalculatorOperation.Divide
    | _ -> raise (System.InvalidOperationException("Incorrect operation"))
    
let parseArgument (arg: string) =
    match System.Double.TryParse(arg) with
    | (true, num) -> num
    | _ -> raise (System.ArgumentException("It not number!"))
    
let parseCalcArguments (args: string[]) =
    if (not (isArgLengthSupported args)) then
        raise (System.ArgumentException("Incorrect number of arguments"))
    let args = {
        arg1 = parseArgument args.[0]
        arg2 = parseArgument args.[2]
        operation = parseOperation args.[1] 
    }
    args
